extensions/parquet/table/src/main/java/io/deephaven/parquet/table/location/ParquetTableLocation.java

```diff
@@ -713,79 +719,158 @@ public void pushdownFilter(
+                    }
+                    break;
+                }
+                case DeferredDataIndex: {
+                    final BasicDataIndex dataIndex =
+                            hasDataIndex(parquetColumnNames) ? getDataIndex(parquetColumnNames) : null;
+                    if (dataIndex == null) {
+                        continue;
+                    }
+                    try (final PushdownResult ignored = result) {
+                        result = pushdownDataIndex(selection, filter, renameMap, dataIndex, result);
+                    }
+                    break;
+                }
+                default:
+                    throw new IllegalStateException("Unexpected value: " + mode);
             }
             if (result.maybeMatch().isEmpty()) {
                 // No maybe rows remaining, so no reason to continue filtering.
-                onComplete.accept(result);
-                return;
+                break;
             }
         }
+        onComplete.accept(result);
+    }
 
-        // Pushdown via dictionary
-        if (shouldExecute(QueryTable.DISABLE_WHERE_PUSHDOWN_PARQUET_DICTIONARY,
-                PushdownResult.DICTIONARY_DATA_COST, executedFilterCost, costCeiling)
-                && ctx.supportsChunkFilter()
-                && hasDictionaryPage(parquetColumnNames[0], ctx.columnDefinitions().get(0))) {
-            try (final PushdownResult ignored = result) {
-                result = pushdownFilterDictionary(selection, ctx, parquetColumnNames, result);
-            }
-            if (result.maybeMatch().isEmpty()) {
-                // No maybe rows remaining, so no reason to continue filtering.
-                onComplete.accept(result);
-                return;
-            }
+    // ---------------------------------------------------------------------------------------------------------------
+    // The following should be _cheap_ checks that don't require materializing Parquet metadata to check.
+    // ---------------------------------------------------------------------------------------------------------------
+    // Note: in the future, we this would be an easy way to allow turning off Parquet pushdown on a location by location
+    // basis; we could expose the options through ParquetInstructions
+
+    private boolean supportsMetadataFiltering() {
+        return true;
+    }
+
+    private boolean supportsDictionaryFiltering() {
+        return true;
+    }
+
+    private boolean supportsInMemoryDataIndexFiltering() {
+        return hasAnyCachedDataIndex();
+    }
+
+    private boolean supportsDeferredDataIndexFiltering() {
+        return true;
+    }
+
+    // ---------------------------------------------------------------------------------------------------------------
+
+    enum PushdownMode {
+        // @formatter:off
+        ParquetRowGroupMetadata(
+                QueryTable.DISABLE_WHERE_PUSHDOWN_PARQUET_ROW_GROUP_METADATA,
```

Comment:
Java question: I expect this enum is populated once, which locks in the `disabled` flag and will not reflect dynamic user changes. In performance testing I enable/disable these flags.

If we want to stay dynamic, could make this a lambda to return the current state of the disable flag.

Or am I wrong and this enum is continually created/destroyed?

Reply:
Good catch. Although I'm not very happy w/ non-final static values, best to retain the current behavior until it can be improved in the future.